AddAdmin/RemoveAdmin calls in Communities service

schradert opened the pull request:
Currently imports some functions from `admin.proto` and `admin.py` servicer because they were originally implemented there and they are expected to return user details rather than a different response payload

<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->

Closes #4295 

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on the web frontend)
If you need help with any of these, please ask :)
--->
**Backend checklist**
- [x] Formatted my code by running `autoflake --exclude src/proto -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works

<!---
Remember to request review from couchers-org/web, couchers-org/backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

schradert committed:
feat: AddAdmin/RemoveAdmin calls on Communities

Currently imports some functions from `admin.proto` and `admin.py`
servicer because they were originally implemented there and they are
expected to return user details rather than a different response payload

schradert committed:
chore: autoflake formatting

schradert committed:
fix: recursive protocol buffer import

this is done by simplifying the response of AddAdmin and RemoveAdmin to
just include the community_id passed into the request, exactly like
JoinCommunity and LeaveCommunity work

schradert commented:
Question for @aapeliv: how can we make sure that only admins of an ancestor node (or even just world admin) can use these endpoints?

schradert commented:
Currently having annoying issues with local podman I need to debug tomorrow in order to run the database properly and confirm tests work on my side

aapeliv commented on app/proto/communities.proto:
```diff
@@ -30,6 +30,14 @@ service Communities {
+  rpc RemoveAdmin(RemoveAdminReq) returns (RemoveAdminRes) {
+    // Removes user as community builder
+  }
+
```
I think maybe better to retun `google.protobuf.Empty`?

aapeliv commented on app/backend/src/couchers/servicers/communities.py:
```diff
@@ -180,6 +180,43 @@ def ListAdmins(self, request, context):
 
+    def AddAdmin(self, request, context):
+        with session_scope() as session:
+            user = session.execute(select(User).where_username_or_email_or_id(request.user)).scalar_one_or_none()
+            if not user:
+                context.abort(grpc.StatusCode.NOT_FOUND, errors.USER_NOT_FOUND)
```
I think better to pass in the `user_id` and do something like

```py
session.execute(select(User).where_users_visible(context).where(User.id == request.user_id)).scalar_one_or_none()
```

(The `where_[...]_email` function is only for admins, otherwise emails are private).

aapeliv commented on app/backend/src/couchers/servicers/communities.py:
```diff
@@ -180,6 +180,43 @@ def ListAdmins(self, request, context):
                 next_page_token=str(admins[-1].id) if len(admins) > page_size else None,
             )
 
+    def AddAdmin(self, request, context):
```
I think you should do the permission check using `can_moderate_node(session, context.user_id, node.id)` (where `node.id == request.community_id`, but you probably need to grab the node first.

aapeliv reviewed (changes requested):
Looks good, please see the comments!

schradert committed:
fix: make Add/Remove Admin response empty

documented FAILED_PRECONDITION situations in protocol buffer to more
closely resemble the existing JoinCommunity and LeaveCommunity endpoints

schradert committed:
improve: only handle user_id for Add/Remove Admin

emails are private except for world admins, so
`where_username_or_email_or_id` should only be in the admin api calls

schradert committed:
feat: node moderation permission for admin mgmt

schradert committed:
chore: autoflake

schradert commented:
@aapeliv why do you think this error is showing up in the tests:

```
    def do_execute(self, cursor, statement, parameters, context=None):
>       cursor.execute(statement, parameters)
E       sqlalchemy.exc.IntegrityError: (psycopg2.errors.ForeignKeyViolation) insert or update on table "cluster_subscriptions" violates foreign key constraint "fk_cluster_subscriptions_cluster_id_clusters"
E       DETAIL:  Key (cluster_id)=(4) is not present in table "clusters".
E       
E       [SQL: INSERT INTO cluster_subscriptions (user_id, cluster_id, role) VALUES (%(user_id)s, %(cluster_id)s, %(role)s) RETURNING cluster_subscriptions.id]
E       [parameters: {'user_id': 3, 'cluster_id': 4, 'role': 'member'}]
E       (Background on this error at: https://sqlalche.me/e/20/gkpj)
```

It was able to get ID 4 in the first place on the line `c1_id = get_community_id(session, "Country 1")`, so I'm confused why it doesn't exist now...

aapeliv commented on app/backend/src/couchers/servicers/communities.py:
```diff
@@ -180,6 +180,59 @@ def ListAdmins(self, request, context):
+                context.abort(grpc.StatusCode.NOT_FOUND, errors.USER_NOT_FOUND)
+
+            subscription = user.cluster_subscriptions.where(
+                ClusterSubscription.cluster_id == request.community_id
```
Cluster id and node id are different. I think here you need `node.official_cluster.id`

schradert replied:
Oh I see, I wasn't paying close enough attention to the fact that `community_id` is a node id. I should've noticed that when just adding the `node` logic I found in the other functions.

aapeliv commented:
> @aapeliv why do you think this error is showing up in the tests:

See that comment I just added. If that wasn't the issue then please quote the line number/function where this is raised.

schradert committed:
fix: treat community_id as node_id, not cluster_id

the tests also don't create new data erroneously but use fixture data
that fit the testing requirements better

schradert committed:
fix: find matching subscription outside query

Tests were failing with a `InstrumentedList has no attribute "where"`
because `user` has already been evaluated against the database, so we
need to match (defaulting to None if no match) in standard python. Given
that ClusterSubscription is unique by `(user_id, cluster_id)`, this
`next` solution is guaranteed to return one or zero elements.

aapeliv commented on app/backend/src/couchers/errors.py:
```diff
@@ -96,6 +96,7 @@
 NO_PASSWORD = "That account has no password."
 NO_PENDING_VERIFICATION = "No phone number to verify."
 NO_RECIPIENTS = "No recipients."
+NODE_MODERATE_PERMISSION_DENIED = "You're not allowed to moderate that node"
```
Maybe replace node with community in user facing error?

aapeliv commented on app/backend/src/couchers/servicers/communities.py:
```diff
@@ -194,9 +194,9 @@ def AddAdmin(self, request, context):
-            subscription = user.cluster_subscriptions.where(
-                ClusterSubscription.cluster_id == node.official_cluster.id
-            ).one_or_none()
+            subscription = next(
+                (sub for sub in user.cluster_subscriptions if sub.cluster_id == node.official_cluster.id), None
+            )
```
Sorry. I think better to replace with something like
```
            subscription = session.execute(
                select(ClusterSubscription)
                .where(ClusterSubscription.user_id == user.id)
                .where(ClusterSubscription.cluster_id == node.official_cluster.id)
            ).scalar_one_or_none()
```

aapeliv commented on app/proto/communities.proto:
```diff
@@ -151,6 +168,16 @@ message ListAdminsRes {
 }
 
+message AddAdminReq {
+  uint32 user_id = 1;
```
`user_id` should be `int64`

schradert replied:
Sorry, I see that in UserDetails now

aapeliv reviewed (changes requested):
Just the three small comments. Sorry about the `.where` issue with `cluster_subscriptions`, not sure why that didn't work!

schradert committed:
improve: clearer user facing error for moderation

schradert committed:
improve: replace next with query for subscription

schradert committed:
fix: user_id is int64

aapeliv reviewed (dismissed):
Looks good to me now, so when whatever checks are failing are sorted it’s good to go!

schradert commented on app/backend/src/tests/test_communities.py:
```diff
@@ -563,6 +563,72 @@ def test_ListAdmins(testing_communities):
+
+        with communities_session(token4) as api:
+            res = api.ListAdmins(communities_pb2.ListAdminsReq(community_id=node_id))
+            assert res.admin_user_ids == [user4_id, user5_id]
```
The test is only failing here in CI because it expects user 2 to be an admin already, but within the context of `test_RemoveAdmin` this doesn't make any sense. When this test is run in isolation, it succeeds because the preceding attempt to add user 2 AS user 8 fails as expected. Could it be failing because of incomplete cleanup from `test_AddAdmin` where the final list of admin ids does include user 2?

aapeliv replied:
Yes, I think maybe the communities don’t clean up the whole thing but reuse between tests.

Maybe you can make/use another community to make sure they are separate?

schradert replied:
I just added a line of cleanup to both

aapeliv replied:
So I don't remember the specifics of why the communities tests do one set up and one cleanup. Not sure if it was very slow or something. Sorry about that.

schradert committed:
test(communities/admin): custom rollback
